File: ui/background.py
import threading
import logging
import asyncio
import sys
import subprocess
import os
import time
from nicegui import ui
from config_system import config
from utils import is_port_active
from zena_mode.gateway_telegram import run_gateway as run_telegram_gateway
from zena_mode.gateway_whatsapp import run_whatsapp_gateway
from zena_mode.auto_updater import check_for_updates, get_local_version, ModelScout

logger = logging.getLogger("ZenAI.Background")


def start_backend_server():
    """Start the backend orchestrator if it's not already running."""
    # Prevent infinite recursion: If this Zena instance was launched BY the backend (via ZENA_SKIP_PRUNE),
    # we shouldn't try to launch it again unless checking strictly by port.

    mgmt_port = config.mgmt_port
    if is_port_active(mgmt_port):
        logger.info(f"[Background] Backend Hub already active on port {mgmt_port}")
        return

    logger.info(f"[Background] Starting Backend Orchestrator on port {mgmt_port}...")

    # Run server.py as a separate process
    # We pass --no-ui to prevent the server from trying to launch a *second* UI
    cmd = [sys.executable, "-m", "zena_mode.server", "--no-ui", "--guard-bypass"]

    # If we are in mock mode, we might want to signal that to the server (if it supports it)
    # For now, we just run the standard server.

    try:
        subprocess.Popen(
            cmd,
            cwd=str(config.BASE_DIR),
            creationflags=subprocess.CREATE_NEW_CONSOLE if os.name == "nt" else 0,
            shell=False,
        )
        logger.info("[Background] Backend launch command issued.")

        # Give it a second to bind
        time.sleep(2)
        if not is_port_active(mgmt_port):
            logger.warning("[Background] Backend port still closed after simple launch.")

    except Exception as e:
        logger.error(f"[Background] Failed to launch backend: {e}")


def start_background_gateways():
    """Launch backend services and messaging gateways."""

    # 1. Ensure Backend API (Orchestrator) is running
    # This is critical for Council, Voice, and System features that use the API
    if config.system.auto_start_backend:
        threading.Thread(target=start_backend_server, daemon=True, name="BackendLauncher").start()

    # 2. Gateways
    zena_config = config.get("zena_mode", {})
    if zena_config.get("telegram_token"):
        threading.Thread(target=run_telegram_gateway, daemon=True, name="TelegramBot").start()

    if zena_config.get("whatsapp_whitelist"):
        threading.Thread(target=run_whatsapp_gateway, daemon=True, name="WhatsAppBot").start()


async def run_system_checks():
    """Perform background auto-update and model scout checks."""
    zena_config = config.get("zena_mode", {})
    if zena_config.get("auto_update_enabled", True):
        local_tag = await get_local_version()
        update_info = check_for_updates(current_tag=local_tag)
        if update_info:
            ui.notify(f"💎 Shiny new update available: {update_info['tag']}!", color="info", position="bottom-right")

        # Scouting for "Best in Class" models
        scout = ModelScout()
        coding_models = scout.find_shiny_models("coding")
        if coding_models:
            logger.info(f"[ModelScout] Found {len(coding_models)} shiny models on HF.")

File: rust_output/src/background.rs
use anyhow::{Result, Context};
use crate::auto_updater::{check_for_updates, get_local_version, ModelScout};
use crate::config_system::{config};
use crate::gateway_telegram::{run_gateway};
use crate::gateway_whatsapp::{run_whatsapp_gateway};
use crate::utils::{is_port_active};
use std::collections::HashMap;
use tokio;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Start the backend orchestrator if it's not already running.
pub fn start_backend_server() -> Result<()> {
    // Start the backend orchestrator if it's not already running.
    let mut mgmt_port = config::mgmt_port;
    if is_port_active(mgmt_port) {
        logger.info(format!("[Background] Backend Hub already active on port {}", mgmt_port));
        return;
    }
    logger.info(format!("[Background] Starting Backend Orchestrator on port {}...", mgmt_port));
    let mut cmd = vec![sys::executable, "-m".to_string(), "zena_mode.server".to_string(), "--no-ui".to_string(), "--guard-bypass".to_string()];
    // try:
    {
        subprocess::Popen(cmd, /* cwd= */ config::BASE_DIR.to_string(), /* creationflags= */ if os::name == "nt".to_string() { subprocess::CREATE_NEW_CONSOLE } else { 0 }, /* shell= */ false);
        logger.info("[Background] Backend launch command issued.".to_string());
        std::thread::sleep(std::time::Duration::from_secs_f64(2));
        if !is_port_active(mgmt_port) {
            logger.warning("[Background] Backend port still closed after simple launch.".to_string());
        }
    }
    // except Exception as e:
}

/// Launch backend services and messaging gateways.
pub fn start_background_gateways() -> () {
    // Launch backend services and messaging gateways.
    if config::system.auto_start_backend {
        std::thread::spawn(|| {});
    }
    let mut zena_config = config::get(&"zena_mode".to_string()).cloned().unwrap_or(HashMap::new());
    if zena_config.get(&"telegram_token".to_string()).cloned() {
        std::thread::spawn(|| {});
    }
    if zena_config.get(&"whatsapp_whitelist".to_string()).cloned() {
        std::thread::spawn(|| {});
    }
}

/// Perform background auto-update and model scout checks.
pub async fn run_system_checks() -> () {
    // Perform background auto-update and model scout checks.
    let mut zena_config = config::get(&"zena_mode".to_string()).cloned().unwrap_or(HashMap::new());
    if zena_config.get(&"auto_update_enabled".to_string()).cloned().unwrap_or(true) {
        let mut local_tag = get_local_version().await;
        let mut update_info = check_for_updates(/* current_tag= */ local_tag);
        if update_info {
            ui.notify(format!("💎 Shiny new update available: {}!", update_info["tag".to_string()]), /* color= */ "info".to_string(), /* position= */ "bottom-right".to_string());
        }
        let mut scout = ModelScout();
        let mut coding_models = scout.find_shiny_models("coding".to_string());
        if coding_models {
            logger.info(format!("[ModelScout] Found {} shiny models on HF.", coding_models.len()));
        }
    }
}
